tokio-stream/src/stream_ext/timeout_repeating.rs

```diff
@@ -0,0 +1,66 @@
+use crate::stream_ext::Fuse;
+use crate::{Elapsed, Stream};
+use tokio::time::{Instant, Sleep};
+
+use core::future::Future;
+use core::pin::Pin;
+use core::task::{Context, Poll};
+use pin_project_lite::pin_project;
+use std::time::Duration;
+
+pin_project! {
+    /// Stream returned by the [`timeout_repeating`](super::StreamExt::timeout_repeating) method.
+    #[must_use = "streams do nothing unless polled"]
+    #[derive(Debug)]
+    pub struct TimeoutRepeating<S> {
+        #[pin]
+        stream: Fuse<S>,
+        #[pin]
+        deadline: Sleep,
+        duration: Duration,
+    }
+}
+
+impl<S: Stream> TimeoutRepeating<S> {
+    pub(super) fn new(stream: S, duration: Duration) -> Self {
+        let next = Instant::now() + duration;
+        let deadline = tokio::time::sleep_until(next);
+
+        TimeoutRepeating {
+            stream: Fuse::new(stream),
+            deadline,
+            duration,
+        }
+    }
+}
+
+impl<S: Stream> Stream for TimeoutRepeating<S> {
+    type Item = Result<S::Item, Elapsed>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut me = self.project();
+
+        match me.stream.poll_next(cx) {
+            Poll::Ready(v) => {
+                if v.is_some() {
+                    let next = Instant::now() + *me.duration;
+                    me.deadline.reset(next);
+                }
+                return Poll::Ready(v.map(Ok));
+            }
+            Poll::Pending => {}
+        };
+
+        ready!(me.deadline.as_mut().poll(cx));
+        let next = Instant::now() + *me.duration;
+        me.deadline.reset(next);
+        Poll::Ready(Some(Err(Elapsed::new())))
```

Comment:
This will not yield the errors at a regular rate. The duration between two errors will be increased by the time between the timeout triggering and `poll_next` being called. Perhaps we should use an `Interval` instead?

Reply:
I changed it to use an `Interval`. The infrequent polling problem has another edge case: if set to duration `x`, and the underlying stream doesn't produce a value until `x + t1`, but the stream is not polled until `x + t2`, then when polled would it be expected to produce a timeout or the value from the underlying stream?

Reply:
It probably makes sense to add a [`MissedTickBehavior`][1] argument.

> The infrequent polling problem has another edge case: if set to duration `x`, and the underlying stream doesn't produce a value until `x + t1`, but the stream is not polled until `x + t2`, then when polled would it be expected to produce a timeout or the value from the underlying stream?

I'm not sure. What do you think? Maybe it should depend on the `MissedTickBehavior`.

[1]: https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html

Reply:
For my use case, it isn't sensitive to these kinds of timing or ordering issues, but of course I would like this to be as generally useful as possible.

Returning a timeout when there is an actual value waiting to be returned feels a little pedantic but would arguably be the most correct behavior if not the most useful. I lean towards favoring the actual value, or making this a configurable behavior.